File: PythonSandbox/src/leetcode/lc208_implement_trie.py
"""
https://leetcode.com/problems/implement-trie-prefix-tree/description/ 
"""
class Trie:

    def __init__(self):
        self.trie = {}

    def insert(self, word: str) -> None:
        n = self.trie

        for _,c in enumerate(word):
            if c in n.keys():
                n = n[c]
            else:
                n[c] = {}
                n = n[c]
        
        n["*"] = {}

    def search(self, word: str) -> bool:
        n = self.trie
        for _,c in enumerate(word):
            if c in n.keys():
                n = n[c]
            else:
                return False

        if "*" in n.keys() or n == {}:
            return True
        return False

    def startsWith(self, prefix: str) -> bool:
        n = self.trie
        for _,c in enumerate(prefix):
            if c in n.keys():
                n = n[c]
            else:
                return False
        return True

File: JavaSandbox/src/leetcode/LC208_ImplementTrie.java
package leetcode;
// https://leetcode.com/problems/implement-trie-prefix-tree/description/

import java.util.*;

public class LC208_ImplementTrie {
    HashMap<Character, HashMap> trie;

    public LC208_ImplementTrie() {
        trie = new HashMap<Character, HashMap>();
    }
    
    public void insert(String word) {
        HashMap<Character, HashMap> n = trie;
        for(Character currChar: word.toCharArray()) {
            if(n.keySet().contains(currChar)) {
                n = n.get(currChar);
            } else {
                n.put(currChar, new HashMap<Character, HashMap>());
                n = n.get(currChar);
            }
        }
        n.put('*', new HashMap<Character, HashMap>());
    }
    
    public boolean search(String word) {
        HashMap<Character, HashMap> n = trie;
        for(Character currChar: word.toCharArray()) {
            if(n.keySet().contains(currChar)) {
                n = n.get(currChar);
            } else {
                return false;
            }
        }

        if(n.keySet().contains('*') || n.isEmpty()) {
            return true;
        }
        return false;
    }
    
    public boolean startsWith(String prefix) {
        HashMap<Character, HashMap> n = trie;
        for(Character currChar: prefix.toCharArray()) {
            if(n.keySet().contains(currChar)) {
                n = n.get(currChar);
            } else {
                return false;
            }
        }
        return true;
    }
}
